node/bft/src/helpers/cache_round.rs

```diff
@@ -0,0 +1,248 @@
+            let old_round = self.address_rounds[address_index].1;
+            self.address_rounds.remove(address_index);
+            self.prune_validator_from_highest_rounds(old_round, address_x)?;
+        }
+        Ok(())
+    }
+
+    /// Update based on a new (round, address) pair seen in the wild. This does two things:
+    /// - If the round is higher than a previous one from this address, set it in highest_rounds
+    /// - Keep incrementing `last_highest_round_with_quorum` as long as it passes a stake-weighted quorum
+    /// We ignore the case where tomorrow's stake-weighted quorum round is *lower* than the current one
+    pub fn update(&mut self, round: u64, validator_address: Address<N>, committee: &Committee<N>) -> Result<u64> {
+        ensure!(committee.members().contains_key(&validator_address), "Address is not a member of the committee");
+        let validator = AddressWithCoordinate::from(validator_address);
+
+        let mut inserted = false;
+        // Only consider updating if we see a high round
+        if round > self.last_highest_round_with_quorum {
+            match self.address_rounds.binary_search_by_key(&validator.x, |&(a, _)| a.x) {
+                // We recognized the validator, so we may have to update it
+                Ok(address_index) => {
+                    let (_, old_round) = self.address_rounds[address_index];
+                    // Should we update the validator's highest seen round?
+                    if old_round < round {
+                        inserted = true;
+                        self.address_rounds[address_index].1 = round;
+                        self.prune_validator_from_highest_rounds(old_round, validator.x)?;
+                        self.insert_validator_at_round(round, validator);
+                    }
+                }
+                // We did not recognize the validator, so we should add it
+                Err(address_index) => {
+                    inserted = true;
+                    self.address_rounds.insert(address_index, (validator, round));
+                    self.insert_validator_at_round(round, validator);
+                }
+            }
+            // If we cached more validators than the current committee size, we should prune
+            if self.address_rounds.len() > committee.num_members() {
+                self.prune_stale_validators(committee)?;
+            }
+            // Confirm we did not cache more validators than the current committee size
+            ensure!(self.address_rounds.len() <= committee.num_members());
+            // Confirm we did not cache more validators than the current committee size
+            ensure!(self.highest_rounds.iter().map(|(_, a)| a.len()).sum::<usize>() <= committee.num_members());
+        }
+
+        // Check if we reached quorum on a new round
+        if inserted {
+            while committee.is_quorum_threshold_reached(&self.validators_in_support(committee)?) {
```

Comment:
The committee being used here may changed based on the round you are checking.

Reply:
I think its fine to use the single given committee (from `get_previous_committee_for_round`) to update completely, because who cares whether old outdated committees have high rounds..

Sidenote: if we were to only update by 1 round at a time based on the given committee, instead of always taking just `<quorum>` batch proposals to update to the correct round, we will then need `<quorum> + i` batch proposals to increase our round by `i`.  If `<quorum> > max_gc_rounds`, then we might never catch up...

Reply:
Is there any case where you won't ever be able to see quorum because the fixed committee you are using doesn't include newly bonded validators?

Reply:
Great question, it seems so yes... Because we use our old outdated round to determine the committee: `self.ledger.get_previous_committee_for_round(self.current_round())`.

The only alternative I see now is to use the peer's advertised `batch_round` - but one question on that: is it possible that we don't know the committee yet for rounds far in the future so that `get_previous_committee_for_round(batch_round)?` will keep failing? Or do we expect that to succeed at some point?